File: 72-edit-distance/edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.length();
        int n2 = word2.length();
        int mat[n1+1][n2+1];
        for(int i=0; i<n2+1; i++){
            mat[0][i] = i;
        }
        for(int i=0; i<n1+1; i++){
            mat[i][0] = i;
        }
        for(int i=1; i<n1+1; i++){
            for(int j=1; j<n2+1; j++){
                if(word1[i-1] == word2[j-1]) mat[i][j] = mat[i-1][j-1];
                else mat[i][j] = min(1 + mat[i-1][j-1], min(mat[i][j-1] + 1, mat[i-1][j] + 1));
            }
        }

        return mat[n1][n2];
    }
};

File: 72-edit-distance/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:

        if len(word1) < len(word2):
            self.minDistance(word2, word1)

        dp = [[-1]*len(word2) for _ in range(len(word1))]

        def recur(i, j):
            if i == len(word1) and j == len(word2):
                return 0
            elif i == len(word1):
                return len(word2)-j
            elif j == len(word2):
                return len(word1)-i

            if dp[i][j] != -1:
                return dp[i][j]
            
            if word1[i] == word2[j]:
                dp[i][j] = recur(i+1, j+1)
            else:
                dp[i][j] = 1 + min(recur(i, j+1), recur(i+1, j), recur(i+1, j+1))
            
            return dp[i][j]
        
        return recur(0, 0)

        